### ISSUE TITLE
#lib kills the build

### ISSUE THREAD

**Issue by dfreelan** [bytedeco/javacpp-presets#217]
This change:
`make -j $MAKEJ BLAS=$BLAS OPENCV_VERSION=3 DISTRIBUTE_DIR=.. # lib`

Makes it so i cannot install on my linux x86_64 machine.

I get an error like this:

```

CXX/LD -o .build_release/tools/device_query.bin
CXX/LD -o .build_release/tools/extract_features.bin
CXX/LD -o .build_release/tools/test_net.bin
CXX/LD -o .build_release/tools/train_net.bin
.build_release/lib/libcaffe.so: undefined reference to `compress2'
.build_release/lib/libcaffe.so: undefined reference to `inflate'
.build_release/lib/libcaffe.so: undefined reference to `inflateInit_'
.build_release/lib/libcaffe.so: undefined reference to `inflateEnd'
collect2: error: ld returned 1 exit status
make: *** [.build_release/tools/train_net.bin] Error 1
make: *** Waiting for unfinished jobs....
.build_release/lib/libcaffe.so: undefined reference to `compress2'
.build_release/lib/libcaffe.so: undefined reference to `inflate'
.build_release/lib/libcaffe.so: undefined reference to `inflateInit_'
.build_release/lib/libcaffe.so: undefined reference to `inflateEnd'
collect2: error: ld returned 1 exit status
make: *** [.build_release/tools/extract_features.bin] Error 1
.build_release/lib/libcaffe.so: undefined reference to `compress2'
.build_release/lib/libcaffe.so: undefined reference to `inflate'
.build_release/lib/libcaffe.so: undefined reference to `inflateInit_'
.build_release/lib/libcaffe.so: undefined reference to `inflateEnd'
collect2: error: ld returned 1 exit status
make: *** [.build_release/tools/test_net.bin] Error 1

```
I honestly can't say I really know what's going on here at all.... but i somehow managed to figure that out!

idk if the wiki needs to updated or what? But I spent a good part of a day trying to find out what the hel was going wrong!

**Comment by saudet** [comment]
So, if I get this right, the previous version of the line as below works on your machine??
```
make -j $MAKEJ BLAS=$BLAS OPENCV_VERSION=3 DISTRIBUTE_DIR=.. lib
```

**Comment by saudet** [comment]
In any case, if you're looking into making builds for Linux, please try to follow these instructions:
https://github.com/bytedeco/javacpp-presets/wiki/Build-Environments#linux-x86-and-x86_64

**Comment by dfreelan** [comment]
Yes that is the instruction set i'm using 

If I follow that wiki naiively, I end up with errors for projects i don't care about (e.g. flycatcher)

So I modify it by  just installing caffe as follows:
bash cppbuild.sh -platform linux-x86_64 install caffe

but now when i execute (notice it is also modified to just be caffe)
mvn clean install -projects caffe -Djavacpp.platform=linux-x86_64 -Djavacpp.platform.dependency=false

I get the following error:

`[ERROR] Failed to execute JavaCPP Builder: /home/dfreelan/dev/javacppSources/javacpp-presets/caffe/cppbuild/linux-x86_64/include/google/protobuf/stubs/common.h:373: Unexpected token '>'
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.524s
[INFO] Finished at: Thu May 12 01:31:37 EDT 2016
[INFO] Final Memory: 39M/1232M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.bytedeco:javacpp:1.1:build (generate-sources) on project caffe: Failed to execute JavaCPP Builder: /home/dfreelan/dev/javacppSources/javacpp-presets/caffe/cppbuild/linux-x86_64/include/google/protobuf/stubs/common.h:373: Unexpected token '>' -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
`

**Comment by saudet** [comment]
Sounds like you're trying to use an old version of JavaCPP.

In any case, if you get these kind of errors in Docker, you should file bug upstream. It's not something that should happen in Docker.

**Comment by dfreelan** [comment]
I assumed we had to compile it outside of Docker, as I couldn't get to the directory where  I cloned javaCPP and javacpp-presets

Do I have to clone them again inside the docker? or is there some docker specific way to navigate to my home directory (while in docker, `cd /home/<my_user_name>/dev` returns `bash: cd: /home/dfreelan/dev: No such file or directory
`

**Comment by dfreelan** [comment]
Looks like you're correct about me trying to use an old version.


I THOUGHT I had `mvn clean installed` in my freshly pulled `javacpp`, but i was obviously mistaken!

Thanks! that's gonna save me a lot of time.

**Comment by dfreelan** [comment]
I redact the previous statement... It just looked like it was making progress because I didn't specify the caffe project :(


I execute the following:
`rm -rf javacpp
git clone https://github.com/bytedeco/javacpp.git
cd javacpp
mvn clean install
bash cppbuild.sh -platform linux-x86_64 install caffe
mvn clean install -projects caffe -Djavacpp.platform=linux-x86_64 -Djavacpp.platform.dependency=false
`


And I still get the same thing. But i'm not in docker :/


The tutorial seems to be pretty straight forward, except for how to get to your javacpp-presets folder from docker.

**Comment by saudet** [comment]
Well, just run git clone in it, yes, that works.

**Comment by dfreelan** [comment]
I totally erased my docker, made a new one, installed everything from scratch using the same sort of commands.


The modifications I had to make was to the same file, to get cppbuild.sh to work for caffe, and I also had to tell cppbuild.sh to only build caffe instead of EVERYTHING (since that includes things like flycatcher)

now I'm getting a different error:

`[ERROR] Failed to execute JavaCPP Builder: /home/dfreelan/dev/javacppSources/javacpp-presets/caffe/cppbuild/linux-x86_64/include/google/protobuf/stubs/common.h:373: Unexpected token '>'
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.576s
[INFO] Finished at: Thu May 12 03:58:00 EDT 2016
[INFO] Final Memory: 42M/1482M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.bytedeco:javacpp:1.1:build (generate-sources) on project caffe: Failed to execute JavaCPP Builder: /home/dfreelan/dev/javacppSources/javacpp-presets/caffe/cppbuild/linux-x86_64/include/google/protobuf/stubs/common.h:373: Unexpected token '>' -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.bytedeco:javacpp:1.1:build (generate-sources) on project caffe: Failed to execute JavaCPP Builder
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:217)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to execute JavaCPP Builder
	at org.bytedeco.javacpp.tools.BuildMojo.execute(BuildMojo.java:276)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	... 19 more
Caused by: org.bytedeco.javacpp.tools.ParserException: /home/dfreelan/dev/javacppSources/javacpp-presets/caffe/cppbuild/linux-x86_64/include/google/protobuf/stubs/common.h:373: Unexpected token '>'
	at org.bytedeco.javacpp.tools.Token.expect(Token.java:98)
	at org.bytedeco.javacpp.tools.Parser.template(Parser.java:254)
	at org.bytedeco.javacpp.tools.Parser.declarations(Parser.java:2291)
	at org.bytedeco.javacpp.tools.Parser.namespace(Parser.java:2253)
	at org.bytedeco.javacpp.tools.Parser.declarations(Parser.java:2345)
	at org.bytedeco.javacpp.tools.Parser.namespace(Parser.java:2253)
	at org.bytedeco.javacpp.tools.Parser.declarations(Parser.java:2345)
	at org.bytedeco.javacpp.tools.Parser.namespace(Parser.java:2253)
	at org.bytedeco.javacpp.tools.Parser.declarations(Parser.java:2345)
	at org.bytedeco.javacpp.tools.Parser.parse(Parser.java:2421)
	at org.bytedeco.javacpp.tools.Parser.parse(Parser.java:2517)
	at org.bytedeco.javacpp.tools.Builder.parse(Builder.java:67)
	at org.bytedeco.javacpp.tools.Builder.build(Builder.java:624)
	at org.bytedeco.javacpp.tools.BuildMojo.execute(BuildMojo.java:269)
	... 21 more
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
`

I assumed this meant I didn't install javacpp correctly, so i rebuilt it by going into the javacpp directory, and executing 
`mvn clean install`

but to no avail, still the same error after rebuilding everything :/


[ERROR] Failed to execute JavaCPP Builder: /home/dfreelan/dev/javacppSources/javacpp-presets/caffe/cppbuild/linux-x86_64/include/google/protobuf/stubs/common.h:373: Unexpected token '>'
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 5.576s
[INFO] Finished at: Thu May 12 03:58:00 EDT 2016
[INFO] Final Memory: 42M/1482M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.bytedeco:javacpp:1.1:build (generate-sources) on project caffe: Failed to execute JavaCPP Builder: /home/dfreelan/dev/javacppSources/javacpp-presets/caffe/cppbuild/linux-x86_64/include/google/protobuf/stubs/common.h:373: Unexpected token '>' -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.bytedeco:javacpp:1.1:build (generate-sources) on project caffe: Failed to execute JavaCPP Builder
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:217)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to execute JavaCPP Builder
	at org.bytedeco.javacpp.tools.BuildMojo.execute(BuildMojo.java:276)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	... 19 more
Caused by: org.bytedeco.javacpp.tools.ParserException: /home/dfreelan/dev/javacppSources/javacpp-presets/caffe/cppbuild/linux-x86_64/include/google/protobuf/stubs/common.h:373: Unexpected token '>'
	at org.bytedeco.javacpp.tools.Token.expect(Token.java:98)
	at org.bytedeco.javacpp.tools.Parser.template(Parser.java:254)
	at org.bytedeco.javacpp.tools.Parser.declarations(Parser.java:2291)
	at org.bytedeco.javacpp.tools.Parser.namespace(Parser.java:2253)
	at org.bytedeco.javacpp.tools.Parser.declarations(Parser.java:2345)
	at org.bytedeco.javacpp.tools.Parser.namespace(Parser.java:2253)
	at org.bytedeco.javacpp.tools.Parser.declarations(Parser.java:2345)
	at org.bytedeco.javacpp.tools.Parser.namespace(Parser.java:2253)
	at org.bytedeco.javacpp.tools.Parser.declarations(Parser.java:2345)
	at org.bytedeco.javacpp.tools.Parser.parse(Parser.java:2421)
	at org.bytedeco.javacpp.tools.Parser.parse(Parser.java:2517)
	at org.bytedeco.javacpp.tools.Builder.parse(Builder.java:67)
	at org.bytedeco.javacpp.tools.Builder.build(Builder.java:624)
	at org.bytedeco.javacpp.tools.BuildMojo.execute(BuildMojo.java:269)
	... 21 more
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

**Comment by dfreelan** [comment]
as proof I am installing the most recent one.. and not 1.1 like that seems to suggest (I 've only been installing from the repo! and this is a clean docker!)

`[INFO] Installing /home/dfreelan/dev/javacppSources/javacpp/target/javacpp-javadoc.jar to /home/dfreelan/.m2/repository/org/bytedeco/javacpp/1.2-SNAPSHOT/javacpp-1.2-SNAPSHOT-javadoc.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 21.151s
[INFO] Finished at: Thu May 12 04:10:54 EDT 2016
[INFO] Final Memory: 30M/743M
[INFO] ------------------------------------------------------------------------
`


I dont know why it seems to be using 1.1, i dont know how it got the idea to try and use 1.1, as i don't have it!

**Comment by dfreelan** [comment]
It's the POM!!!!

Testing it out before i post it... I will edit this post....

**Comment by saudet** [comment]
Looks like you worked out all the issues you were having. Anyway, up-to-date binaries are also included in the recently released version 1.2, so you could just use those. :)

**Comment by liaocs2008** [comment]
I had the same problem and fixed it by editing Caffe Makefile (as shown below in bold):

> $(DYNAMIC_NAME): $(OBJS) | $(LIB_BUILD_DIR)
        @ echo LD -o $@
        $(Q)$(CXX) -shared -o $@ $(OBJS) $(VERSIONFLAGS) $(LINKFLAGS) $(LDFLAGS) $(DYNAMIC_FLAGS) **-lz**
        @ cd $(BUILD_DIR)/lib; rm -f $(DYNAMIC_NAME_SHORT);   ln -s $(DYNAMIC_VERSIONED_NAME_SHORT) $(DYNAMIC_NAME_SHORT)

> $(STATIC_NAME): $(OBJS) | $(LIB_BUILD_DIR)
        @ echo AR -o $@
        $(Q)ar rcs $@ $(OBJS) -lz
